dozer-sql/expression/src/scalar/string.rs

```diff
@@ -272,6 +272,200 @@ pub(crate) fn evaluate_to_char(
+
+    if args.len() == 2 {
+        validate_arg_type(
+            &args[0],
+            vec![FieldType::String, FieldType::Text],
+            schema,
+            ScalarFunctionType::Substr,
+            0,
+        )?;
+        validate_arg_type(
+            &args[1],
+            vec![
+                FieldType::UInt,
+                FieldType::U128,
+                FieldType::Int,
+                FieldType::I128,
+            ],
+            schema,
+            ScalarFunctionType::Substr,
+            1,
+        )?;
+    } else {
+        validate_arg_type(
+            &args[0],
+            vec![FieldType::String, FieldType::Text],
+            schema,
+            ScalarFunctionType::Substr,
+            0,
+        )?;
+        validate_arg_type(
+            &args[1],
+            vec![
+                FieldType::UInt,
+                FieldType::U128,
+                FieldType::Int,
+                FieldType::I128,
+            ],
+            schema,
+            ScalarFunctionType::Substr,
+            1,
+        )?;
+        validate_arg_type(
+            &args[2],
+            vec![
+                FieldType::UInt,
+                FieldType::U128,
+                FieldType::Int,
+                FieldType::I128,
+            ],
+            schema,
+            ScalarFunctionType::Substr,
+            2,
+        )?;
+    }
+
+    let ret_type = FieldType::String;
+
+    Ok(ExpressionType::new(
+        ret_type,
+        false,
+        dozer_types::types::SourceDefinition::Dynamic,
+        false,
+    ))
+}
+
+pub(crate) fn evaluate_substr(
+    schema: &Schema,
+    arg: &mut Expression,
+    position: &mut Expression,
+    length: &mut Option<Box<Expression>>,
+    record: &Record,
+) -> Result<Field, Error> {
+    let arg_field = arg.evaluate(record, schema)?;
+    let arg_value = arg_field.to_string();
+
+    let position_field = position.evaluate(record, schema)?;
+    let position_result = position_field.to_i128();
+    if position_result.is_none() {
+        return Err(Error::InvalidFunctionArgument {
+            function_name: "SUBSTR".to_string(),
+            argument_index: 1,
+            argument: position_field,
+        });
+    }
+    let position_value = position_result.unwrap();
+
+    let length_value = match length {
+        Some(length_expr) => {
+            let length_field = length_expr.evaluate(record, schema)?;
+            let length_result = length_field.to_i128();
+            if length_result.is_none() {
+                return Err(Error::InvalidFunctionArgument {
+                    function_name: "SUBSTR".to_string(),
+                    argument_index: 2,
+                    argument: length_field,
+                });
+            }
+            length_result.unwrap()
+        }
+        None => arg_value.len() as i128,
+    };
+
+    let mut iter = arg_value.char_indices();
+    let (start, _) = iter.nth(position_value.try_into().unwrap()).unwrap();
```

Comment:
This will panic if the indices fall outside of the string's length. If the start is outside the string's length, we should return an empty string, and if the end falls outside of the string, we should just until the end of the string.

So, something like:
```rust
arg_value[start.unwrap_or(arg.len())..end.unwrap_or(arg.len())]
```